localstripe/resources.py

```diff
@@ -2359,67 +2359,71 @@ def _update(self, metadata=None, items=None, trial_end=None,
         except AssertionError:
             raise UserError(400, 'Bad request')
 
-        if items is None or len(items) != 1:
-            raise UserError(500, 'Not implemented')
+        old_plan = self.plan
+        if items is not None:
+            if len(items) != 1:
+                raise UserError(500, 'Not implemented')
 
-        # If no plan specified in update request, we stay on the current one
-        if not items[0].get('plan', None):
-            items[0]['plan'] = self.plan.id
+            # If no plan specified in update request, we stay on the current
+            # one
+            if not items[0].get('plan', None):
+                items[0]['plan'] = self.plan.id
 
-        # To return 404 if not existant:
-        Plan._api_retrieve(items[0]['plan'])
+            # To return 404 if not existant:
+            Plan._api_retrieve(items[0]['plan'])
+
+            # To return 404 if not existant:
+            if items[0]['tax_rates'] is not None:
+                [TaxRate._api_retrieve(tr) for tr in items[0]['tax_rates']]
+
+            self.quantity = items[0]['quantity']
+
+            item = SubscriptionItem(subscription=self.id,
+                                    plan=items[0]['plan'],
+                                    quantity=items[0]['quantity'],
+                                    tax_rates=items[0]['tax_rates'])
+
+            if (self.items._list[0].plan.id != item.plan.id or
+                    self.items._list[0].quantity != item.quantity or
+                    self.items._list[0].tax_rates != item.tax_rates):
+
+                self.items = List('/v1/subscription_items?subscription=' +
+                                  self.id)
+                self.items._list.append(item)
+
+                # Create unused time pending item.
+                # Get previous invoice for this subscription and customer, and
+                # deduce what is already paid:
+                # TODO: Better not to use limit, but take date into account
+                previous = Invoice._api_list_all(None, customer=self.customer,
+                                                 subscription=self.id,
+                                                 limit=99)
+                for previous_invoice in previous._list:
+                    previous_tax_rates = [tr.id for tr in (
+                        previous_invoice.lines._list[0].tax_rates or [])]
+                    InvoiceItem(amount=- previous_invoice.subtotal,
+                                currency=previous_invoice.currency,
+                                proration=True,
+                                description='Unused time',
+                                tax_rates=previous_tax_rates,
+                                customer=self.customer)
```

Comment:
You moved that code *up*, to gather it, and that's a good thing for readability. But is it possible to keep it *gathered* but *below*, just before `create_an_invoice = ...`? Because creating an invoice is the logical continuation of this code...

Reply:
I re-code that function a bunch of time, and it's kind of hard to keep it readable working and logic.
I think moving that part of the code below, and still gathered, would move checks like `# To return 404 if not existant` too, and would not be a good thing.
To keep code that are a sequence of each other closer, I can do the following:
```patch
--- a/localstripe/resources.py
+++ b/localstripe/resources.py
@@ -2360,6 +2360,7 @@ class Subscription(StripeObject):
             raise UserError(400, 'Bad request')
 
         old_plan = self.plan
+        create_an_invoice = False
         if items is not None:
             if len(items) != 1:
                 raise UserError(500, 'Not implemented')
@@ -2386,7 +2387,11 @@ class Subscription(StripeObject):
                                         plan=items[0]['plan'],
                                         quantity=items[0]['quantity'],
                                         tax_rates=items[0]['tax_rates'])
-                self.items._list.append(item)
+                self.items._list.append(item)  # self.plan changed here
+
+                if (self.plan.interval != old_plan.interval or
+                        self.plan.interval_count != old_plan.interval_count):
+                    self._set_up_plan(Plan._api_retrieve(items[0]['plan']))
 
                 # Create unused time pending item.
                 # Get previous invoice for this subscription and customer, and
@@ -2405,6 +2410,15 @@ class Subscription(StripeObject):
                                 tax_rates=previous_tax_rates,
                                 customer=self.customer)
 
+                # If the subscription is updated to a more expensive plan, an
+                # invoice is not automatically generated. To achieve that, an
+                # invoice has to be manually created using the POST /invoices
+                # route.
+                create_an_invoice = (
+                    self.plan.billing_scheme == 'per_unit' and
+                    (self.plan.interval != old_plan.interval or
+                     self.plan.interval_count != old_plan.interval_count))
+
             elif self.items._list[0].tax_rates != items[0]['tax_rates']:
                 self.items = List('/v1/subscription_items?subscription=' +
                                   self.id)
@@ -2426,16 +2440,6 @@ class Subscription(StripeObject):
         if cancel_at_period_end is not None:
             self.cancel_at_period_end = cancel_at_period_end
 
-        if (self.plan.interval != old_plan.interval or
-                self.plan.interval_count != old_plan.interval_count):
-            self._set_up_plan(Plan._api_retrieve(items[0]['plan']))
-
-        # If the subscription is updated to a more expensive plan, an invoice
-        # is not automatically generated. To achieve that, an invoice has to
-        # be manually created using the POST /invoices route.
-        create_an_invoice = self.plan.billing_scheme == 'per_unit' and (
-            self.plan.interval != old_plan.interval or
-            self.plan.interval_count != old_plan.interval_count)
         if create_an_invoice:
             self._create_invoice(items[0])
```
Would it be better? (I pushed it in a FIXUP)

Reply:
I'm not sure. Maybe your initial version was good.

You have the code freshly in mind → you decide!

Reply:
I prefer the initial version I think, because there were less `if` inside `if`.
I hope the code will be understandable enough for the next person that play with it :crossed_fingers: 
Removing the fixup!